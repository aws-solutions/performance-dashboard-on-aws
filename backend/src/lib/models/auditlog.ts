/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export enum ItemEvent {
    Create = "Create",
    Update = "Update",
    Delete = "Delete",
}

export type ModifiedProperty = {
    property: string;
    oldValue: any;
    newValue: any;
};

export interface AuditLog {
    itemId: string;
    timestamp: Date;
    userId: string;
    event: ItemEvent;
    modifiedProperties?: ModifiedProperty[];
}

// Represents a dynamodb item in the Audit Trail table
export interface AuditLogItem {
    pk: string;
    sk: string;
    type: string;
    userId: string;
    event: ItemEvent;
    modifiedProperties?: ModifiedProperty[];
}

export interface DashboardAuditLog extends AuditLog {
    version: number;
}

// Represents a dynamodb item in the Audit Trail table that
// contains attributes specific to Dashboard audit logs.
export interface DashboardAuditLogItem extends AuditLogItem {
    version: number;
}
