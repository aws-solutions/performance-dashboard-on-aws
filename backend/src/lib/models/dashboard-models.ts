export type Dashboard = {
    id: string,
    name: string,
    topicAreaId: string,
    topicAreaName: string,
    description: string,
    createdBy: string,
};

export type DashboardList = Array<Dashboard>;

export interface DashboardItem {
    pk: string,
    sk: string,
    type: string,
    dashboardName: string,
    topicAreaId: string,
    topicAreaName: string,
    description: string,
    createdBy: string,
};