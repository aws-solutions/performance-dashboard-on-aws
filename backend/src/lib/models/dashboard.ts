/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import { Widget } from "./widget";

export const DASHBOARD_ITEM_TYPE = "Dashboard";

export enum DashboardState {
    Draft = "Draft",
    Published = "Published",
    Archived = "Archived",
    /**
     * @type {string}
     * @deprecated in favor of {@link Draft}
     */
    PublishPending = "PublishPending",
    Inactive = "Inactive",
}

export interface DashboardVersion {
    id: string;
    version: number;
    state: DashboardState;
    friendlyURL?: string;
}

export interface Dashboard {
    id: string;
    name: string;
    version: number;
    parentDashboardId: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    /**
     * @type {any}
     * @deprecated no longer supported
     */
    tableOfContents?: any;
    description: string;
    createdBy: string;
    updatedAt: Date;
    updatedBy?: string;
    submittedBy?: string;
    publishedBy?: string;
    archivedBy?: string;
    deletedBy?: string;
    state: DashboardState;
    releaseNotes?: string;
    widgets?: Array<Widget>;
    friendlyURL?: string;
}

export type DashboardList = Array<Dashboard>;

export interface DashboardItem {
    pk: string;
    sk: string;
    type: string;
    version: number;
    parentDashboardId: string;
    dashboardName: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    tableOfContents?: any;
    description: string;
    createdBy: string;
    updatedAt: string;
    updatedBy?: string;
    submittedBy?: string;
    publishedBy?: string;
    archivedBy?: string;
    deletedBy?: string;
    state: string;
    releaseNotes?: string;
    friendlyURL?: string;
}

// Public representation of a dashboard. Hides some fields
// like createdBy and state, which should not be exposed
// on public endpoints.
export interface PublicDashboard {
    id: string;
    name: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    tableOfContents?: any;
    description: string;
    updatedAt: Date;
    widgets?: Array<Widget>;
    friendlyURL?: string;

    // Filled on a homepage search with dashboard content
    // that matches the search query.
    queryMatches?: Array<string>;
}
