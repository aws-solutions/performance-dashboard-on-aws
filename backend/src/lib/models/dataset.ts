/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export enum SourceType {
    IngestApi = "IngestApi",
    FileUpload = "FileUpload",
}

export enum DatasetSchema {
    None = "None",
    Metrics = "Metrics",
}

export interface Dataset {
    id: string;
    createdBy: string;
    fileName: string;
    s3Key: {
        raw: string;
        json: string;
    };
    updatedAt: Date;
    sourceType: SourceType;
    schema?: DatasetSchema;
}

export type DatasetContent = Array<object>;
export type DatasetList = Array<Dataset>;

export interface DatasetItem {
    pk: string;
    sk: string;
    type: string;
    createdBy: string;
    fileName: string;
    s3Key: {
        raw: string;
        json: string;
    };
    updatedAt: string;
    sourceType: string;
    schema?: string;
}
