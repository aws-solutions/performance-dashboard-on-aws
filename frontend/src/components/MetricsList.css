.text-overflow-ellipsis {
  text-overflow: ellipsis;
}
