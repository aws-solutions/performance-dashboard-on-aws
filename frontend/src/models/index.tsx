/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export type TopicArea = {
    id: string;
    name: string;
    createdBy: string;
    dashboardCount: number;
};

export const TopicAreaSortingCriteria = (a: TopicArea, b: TopicArea) => (a.name > b.name ? 1 : -1);

export type PublicTopicArea = {
    id: string;
    name: string;
    dashboards?: Array<Dashboard | PublicDashboard>;
};

export enum DashboardState {
    Draft = "Draft",
    Published = "Published",
    Archived = "Archived",
    PublishPending = "PublishPending",
    Inactive = "Inactive",
}

export enum SourceType {
    IngestApi = "IngestApi",
    FileUpload = "FileUpload",
}

export enum DatasetType {
    DynamicDataset = "DynamicDataset",
    StaticDataset = "StaticDataset",
    CsvFileUpload = "CsvFileUpload",
    CreateNew = "CreateNew",
}

export type Dashboard = {
    id: string;
    name: string;
    version: number;
    parentDashboardId: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    /**
     * @type {any}
     * @deprecated no longer supported
     */
    tableOfContents?: any;
    description?: string;
    releaseNotes?: string;
    widgets: Array<Widget>;
    state: string;
    updatedAt: Date;
    createdBy: string;
    submittedBy?: string;
    publishedBy?: string;
    archivedBy?: string;
    friendlyURL?: string;
    queryMatches?: Array<string>;
};

export type PublicDashboard = {
    id: string;
    name: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    /**
     * @type {any}
     * @deprecated no longer supported
     */
    tableOfContents?: any;
    description?: string;
    widgets: Array<Widget>;
    updatedAt: Date;
    friendlyURL?: string;
    queryMatches?: Array<string>;
};

export type DashboardVersion = {
    id: string;
    version: number;
    state: DashboardState;
    friendlyURL?: string;
};

export type DashboardAuditLog = {
    itemId: string;
    timestamp: Date;
    version: number;
    userId: string;
    event: "Create" | "Update" | "Delete";
    modifiedProperties?: Array<{
        property: string;
        oldValue: any;
        newValue: any;
    }>;
};

export enum WidgetType {
    Text = "Text",
    Chart = "Chart",
    Table = "Table",
    Image = "Image",
    Metrics = "Metrics",
    Section = "Section",
}

export enum ChartType {
    LineChart = "LineChart",
    ColumnChart = "ColumnChart",
    BarChart = "BarChart",
    PartWholeChart = "PartWholeChart",
    PieChart = "PieChart",
    DonutChart = "DonutChart",
}

export interface Widget {
    id: string;
    name: string;
    widgetType: string;
    order: number;
    updatedAt: Date;
    dashboardId: string;
    content: any;
    showTitle: boolean;
    section?: string;
}

export interface ChartWidget extends Widget {
    content: {
        title: string;
        chartType: ChartType;
        datasetId: string;
        summary: string;
        summaryBelow: boolean;
        datasetType?: DatasetType;
        s3Key: {
            raw: string;
            json: string;
        };
        columnsMetadata: ColumnMetadata[];
        sortByColumn?: string;
        sortByDesc?: boolean;
        horizontalScroll?: boolean;
        stackedChart?: boolean;
        significantDigitLabels: boolean;
        dataLabels: boolean;
        computePercentages: boolean;
        showTotal: boolean;
    };
}

export interface TableWidget extends Widget {
    content: {
        title: string;
        datasetId: string;
        summary: string;
        summaryBelow: boolean;
        datasetType?: DatasetType;
        s3Key: {
            raw: string;
            json: string;
        };
        columnsMetadata: ColumnMetadata[];
        sortByColumn?: string;
        sortByDesc?: boolean;
        significantDigitLabels: boolean;
        displayWithPages: boolean;
    };
}

export interface ImageWidget extends Widget {
    content: {
        title: string;
        imageAltText: string;
        summary?: string;
        summaryBelow: boolean;
        s3Key: {
            raw: string;
        };
        filename: string;
        scalePct?: string;
    };
}

export interface MetricsWidget extends Widget {
    content: {
        title: string;
        datasetId: string;
        oneMetricPerRow: boolean;
        datasetType?: DatasetType;
        significantDigitLabels: boolean;
        metricsCenterAlign: boolean;
        s3Key: {
            raw: string;
            json: string;
        };
    };
}

export interface SectionWidget extends Widget {
    content: {
        title: string;
        summary: string;
        widgetIds?: Array<string>;
        showWithTabs: boolean;
        horizontally?: boolean;
    };
}

export enum DatasetSchema {
    None = "None",
    Metrics = "Metrics",
}

export type Dataset = {
    id: string;
    fileName: string;
    s3Key: {
        raw: string;
        json: string;
    };
    createdBy?: string;
    updatedAt?: Date;
    sourceType?: SourceType;
    schema?: DatasetSchema;
};

export type PublicHomepage = {
    title: string;
    description: string;
    dashboards: Array<PublicDashboard>;
};

export type Homepage = {
    title: string;
    description: string;
    updatedAt: Date;
};

export type PublicSettings = {
    dateTimeFormat: {
        date: string;
        time: string;
    };
    navbarTitle?: string;
    colors: {
        primary: string;
        secondary: string;
    };
    topicAreaLabels?: {
        singular: string;
        plural: string;
    };
    customLogoS3Key?: string;
    customFaviconS3Key?: string;
    contactEmailAddress?: string;
    contactUsContent?: string;
    analyticsTrackingId?: string;
};

export type Settings = {
    publishingGuidance: string;
    dateTimeFormat: {
        date: string;
        time: string;
    };
    updatedAt?: Date;
    navbarTitle?: string;
    colors: {
        primary: string;
        secondary: string;
    };
    topicAreaLabels: {
        singular: string;
        plural: string;
    };
    customLogoS3Key?: string;
    customLogoAltText?: string;
    customFaviconS3Key?: string;
    contactEmailAddress?: string;
    adminContactEmailAddress?: string;
    contactUsContent?: string;
    analyticsTrackingId?: string;
};

export type Metric = {
    title: string;
    value: number;
    changeOverTime?: string;
    percentage?: string;
    currency?: string;
    startDate?: string;
    endDate?: string;
};

// Type for the History object in react-router. Defines the
// location state that is common across all routes.
// Usage in a component: useHistory<LocationState>();
export type LocationState = {
    alert?: {
        type: "success" | "warning" | "info" | "error";
        message: string;
        to?: string;
        linkLabel?: string;
    };
    emails?: string;
    usernames?: Array<string>;
    id?: string;
    metrics?: Array<Metric>;
    metric?: Metric;
    position?: number;
    showTitle?: boolean;
    oneMetricPerRow?: boolean;
    significantDigitLabels?: boolean;
    metricsCenterAlign?: boolean;
    metricTitle?: string;
    origin?: string;
    json?: Array<any>;
    staticDataset?: Dataset;
    redirectUrl?: string;
    crumbLabel?: string;
    datasetType?: DatasetType;
};

export enum UserRoles {
    Admin = "Admin",
    Editor = "Editor",
    /**
     * @type {string}
     * @deprecated no longer supported
     */
    Publisher = "Publisher",
    Public = "Public",
}

export type User = {
    userId: string;
    enabled: boolean;
    userStatus: string;
    sub: string;
    email: string;
    roles: Array<UserRoles>;
    createdAt: Date;
    updatedAt: Date;
};

export enum ColumnDataType {
    Number = "Number",
    Text = "Text",
    Date = "Date",
}

export enum NumberDataType {
    Percentage = "Percentage",
    Currency = "Currency",
    "With thousands separators" = "With thousands separators",
}

export enum CurrencyDataType {
    "Dollar $" = "Dollar $",
    "Euro €" = "Euro €",
    "Pound £" = "Pound £",
}

export interface ColumnMetadata {
    columnName: string;
    dataType?: ColumnDataType;
    numberType?: NumberDataType;
    currencyType?: CurrencyDataType;
    hidden: boolean;
}
