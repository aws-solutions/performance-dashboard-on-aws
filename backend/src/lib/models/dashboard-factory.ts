import { v4 as uuidv4 } from "uuid";
import { Dashboard, DashboardItem } from "./dashboard-models";
import { User } from "./user-models";
import topicareaFactory from "./topicarea-factory";

const DASHBOARD: string = 'Dashboard';

function createNew(name: string, topicAreaId: string, topicAreaName: string, user: User): Dashboard {
  return {
    id: uuidv4(),
    name,
    topicAreaId,
    topicAreaName,
    createdBy: user.userId,
  };
}

/**
 * Converts a Dashboard to a DynamoDB item.
 */
function toItem(dashboard: Dashboard): DashboardItem {
  let item: DashboardItem = {
    pk: itemId(dashboard.id),
    sk: itemId(dashboard.id),
    type: DASHBOARD,
    dashboardName: dashboard.name,
    topicAreaName: dashboard.topicAreaName,
    topicAreaId: topicareaFactory.itemId(dashboard.topicAreaId),
    description: dashboard.description,
    overview: dashboard.overview,
    createdBy: dashboard.createdBy,
  };
  return item;
}

/**
 * Converts a DynamoDB item into a Dashboard object
 */
function fromItem(item: DashboardItem): Dashboard {
  const id = item.pk.substring(10);
  let dashboard: Dashboard = {
    id,
    name: item.dashboardName,
    topicAreaId: item.topicAreaId.substring(10),
    topicAreaName: item.topicAreaName,
    description: item.description,
    overview: item.overview,
    createdBy: item.createdBy,
  }
  return dashboard;
}

function itemId(id: string): string { return `${DASHBOARD}#${id}` }

export default {
  toItem,
  createNew,
  fromItem,
  itemId,
};
