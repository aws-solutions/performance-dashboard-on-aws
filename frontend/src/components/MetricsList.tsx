/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React, { createRef } from "react";
import { Metric } from "../models";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
    faArrowUp,
    faArrowDown,
    faEllipsisV,
    faTrash,
    faGripVertical,
} from "@fortawesome/free-solid-svg-icons";
import Button from "./Button";
import "./MetricsList.css";
import { useTranslation } from "react-i18next";
import DropdownMenu from "./DropdownMenu";
import { MenuItem } from "@reach/menu-button";
import {
    DragDropContext,
    Draggable,
    Droppable,
    DropResult,
    ResponderProvided,
} from "react-beautiful-dnd";

interface Props {
    onClick: Function;
    onEdit?: Function;
    onDelete?: Function;
    onMoveUp?: Function;
    onMoveDown?: Function;
    metrics: Array<Metric>;
    register?: Function;
    defaultChecked?: boolean;
    allowAddMetric: boolean;
    onDrag?: Function;
    onDrop?: Function;
}

function MetricsList(props: Props) {
    const caretUpRefs = props.metrics.map(() => createRef<HTMLButtonElement>());
    const caretDownRefs = props.metrics.map(() => createRef<HTMLButtonElement>());
    const { t } = useTranslation();

    const onDelete = (metric: Metric) => {
        if (props.onDelete) {
            props.onDelete(metric);
        }
    };

    const onEdit = (metric: Metric, position: number) => {
        if (props.onEdit) {
            props.onEdit(metric, position);
        }
    };

    const onMoveDown = (index: number) => {
        if (props.onMoveDown) {
            setNextFocus(index, "Down");
            props.onMoveDown(index);
        }
    };

    const onMoveUp = (index: number) => {
        if (props.onMoveUp) {
            setNextFocus(index, "Up");
            props.onMoveUp(index);
        }
    };

    const setNextFocus = (index: number, direction: "Up" | "Down") => {
        let ref;
        /**
         * When moving a metric up or down, the focus should follow the
         * metric to its new position in the list. Determining the new
         * position depends on whether the metric is moving up or down
         * and if the metric has reached a boundary (i.e. beginning or
         * end of the list), in which case the focus should swap to the
         * opposite caret (careUp vs caretDown).
         *
         * We may not need this logic when we implement drag-n-drop
         * because most of the existing libraries already handle browser
         * focus when moving DOM objects.
         */
        if (direction === "Up") {
            ref = index === 1 ? caretDownRefs[1] : caretUpRefs[index];
        } else {
            const secondLast = props.metrics.length - 2;
            ref = index === secondLast ? caretUpRefs[secondLast] : caretDownRefs[index];
        }

        if (ref.current) {
            ref.current.focus();
        }
    };

    function ActionMenu(metric: Metric) {
        return (
            <DropdownMenu
                buttonText=""
                icon={faEllipsisV}
                variant="unstyled"
                ariaLabel={t("Actions")}
                className="margin-1"
            >
                <MenuItem onSelect={() => {}}>
                    <Button
                        variant="unstyled"
                        ariaLabel={t("DeleteContent", {
                            name: metric.title,
                        })}
                        onClick={() => onDelete(metric)}
                    >
                        <FontAwesomeIcon size="xs" icon={faTrash} className="margin-right-1" />
                        {t("Delete")}
                    </Button>
                </MenuItem>
            </DropdownMenu>
        );
    }

    const onDragEnd = (result: DropResult, provided: ResponderProvided) => {
        const { destination, source } = result;
        if (!destination) {
            return;
        }
        if (destination.droppableId === source.droppableId && destination.index === source.index) {
            return;
        }
        if (props.onDrag) {
            props.onDrag(source.index, destination.index);
        }
    };

    return (
        <div className="display-block" role="group" aria-label={t("Metrics")}>
            <label className="margin-bottom-0 margin-top-2 usa-label text-bold">
                {t("Metrics")}
            </label>
            <p className="margin-top-2px usa-hint">{t("MetricsGuidance")}</p>
            <div className="usa-checkbox margin-bottom-2">
                <input
                    className="usa-checkbox__input"
                    id="oneMetricPerRow"
                    type="checkbox"
                    name="oneMetricPerRow"
                    ref={props.register && props.register()}
                    defaultChecked={props.defaultChecked}
                />
                <label className="usa-checkbox__label" htmlFor="oneMetricPerRow">
                    {t("MetricsOnePerRow")}
                </label>
            </div>
            {props.metrics && props.metrics.length ? (
                <div>
                    <DragDropContext onDragEnd={onDragEnd}>
                        <Droppable droppableId="metrics-list">
                            {(provided) => {
                                return (
                                    <div
                                        role="list"
                                        ref={provided.innerRef}
                                        {...provided.droppableProps}
                                        className="padding-top-2 padding-bottom-2"
                                    >
                                        {props.metrics.map((metric, index) => {
                                            return (
                                                <Draggable
                                                    draggableId={index.toString()}
                                                    index={index}
                                                    key={index}
                                                >
                                                    {(provided, snapshot) => {
                                                        return (
                                                            <div
                                                                {...provided.draggableProps}
                                                                ref={provided.innerRef}
                                                                role="listitem"
                                                                aria-label={(index + 1)?.toString()}
                                                                className={`${
                                                                    snapshot.isDragging
                                                                        ? "usa-focus "
                                                                        : ""
                                                                }border-base-lighter border-1px shadow-1 z-200 radius-lg bg-white grid-row margin-y-1`}
                                                            >
                                                                <div className="grid-row grid-col flex-2 padding-1">
                                                                    <div
                                                                        className="text-base-darker grid-col flex-3 text-center display-flex flex-align-center flex-justify-center margin-left-1"
                                                                        {...provided.dragHandleProps}
                                                                    >
                                                                        <FontAwesomeIcon
                                                                            icon={faGripVertical}
                                                                            size="xs"
                                                                        />
                                                                    </div>
                                                                    <div className="grid-col flex-6 text-center display-flex flex-align-center flex-justify-center font-sans-md margin-left-1">
                                                                        {index + 1}
                                                                    </div>
                                                                    <div className="grid-col flex-4 grid-row flex-column text-center margin-left-1 margin-right-1">
                                                                        <div className="grid-col flex-6">
                                                                            {index > 0 ? (
                                                                                <Button
                                                                                    variant="unstyled"
                                                                                    type="button"
                                                                                    className="margin-top-0-important text-base-darker hover:text-base-darkest active:text-base-darkest"
                                                                                    ariaLabel={t(
                                                                                        "ARIA.MoveUp",
                                                                                        {
                                                                                            metric: metric.title,
                                                                                        },
                                                                                    )}
                                                                                    onClick={() =>
                                                                                        onMoveUp(
                                                                                            index,
                                                                                        )
                                                                                    }
                                                                                    ref={
                                                                                        caretUpRefs[
                                                                                            index
                                                                                        ]
                                                                                    }
                                                                                >
                                                                                    <FontAwesomeIcon
                                                                                        id={`${metric.title}-move-up`}
                                                                                        size="xs"
                                                                                        icon={
                                                                                            faArrowUp
                                                                                        }
                                                                                    />
                                                                                </Button>
                                                                            ) : (
                                                                                <br />
                                                                            )}
                                                                        </div>
                                                                        <div className="grid-col flex-6">
                                                                            {index <
                                                                            props.metrics.length -
                                                                                1 ? (
                                                                                <Button
                                                                                    variant="unstyled"
                                                                                    type="button"
                                                                                    className="margin-top-0-important text-base-darker hover:text-base-darkest active:text-base-darkest"
                                                                                    ariaLabel={t(
                                                                                        "ARIA.MoveDown",
                                                                                        {
                                                                                            metric: metric.title,
                                                                                        },
                                                                                    )}
                                                                                    onClick={() =>
                                                                                        onMoveDown(
                                                                                            index,
                                                                                        )
                                                                                    }
                                                                                    ref={
                                                                                        caretDownRefs[
                                                                                            index
                                                                                        ]
                                                                                    }
                                                                                >
                                                                                    <FontAwesomeIcon
                                                                                        id={`${metric.title}-move-down`}
                                                                                        size="xs"
                                                                                        icon={
                                                                                            faArrowDown
                                                                                        }
                                                                                    />
                                                                                </Button>
                                                                            ) : (
                                                                                <br />
                                                                            )}
                                                                        </div>
                                                                    </div>
                                                                </div>
                                                                <div className="border-base-lighter border-left"></div>
                                                                <div className="grid-col flex-10 grid-row margin-y-1">
                                                                    <div
                                                                        className="grid-col flex-11 usa-tooltip"
                                                                        data-position="bottom"
                                                                        title={metric.title}
                                                                    >
                                                                        <div className="margin-left-1 text-no-wrap overflow-hidden text-overflow-ellipsis">
                                                                            <Button
                                                                                variant="unstyled"
                                                                                type="button"
                                                                                className="margin-left-1 margin-top-0-important text-base-dark hover:text-base-darker active:text-base-darkest text-bold"
                                                                                onClick={() =>
                                                                                    onEdit(
                                                                                        metric,
                                                                                        index,
                                                                                    )
                                                                                }
                                                                                ariaLabel={`Edit ${metric.title}`}
                                                                            >
                                                                                {metric.title}
                                                                            </Button>
                                                                        </div>
                                                                    </div>
                                                                    <div className="grid-col flex-1 text-right">
                                                                        {ActionMenu(metric)}
                                                                    </div>
                                                                </div>
                                                            </div>
                                                        );
                                                    }}
                                                </Draggable>
                                            );
                                        })}
                                        {provided.placeholder}
                                    </div>
                                );
                            }}
                        </Droppable>
                    </DragDropContext>

                    {props.allowAddMetric && (
                        <div className="text-center margin-top-2">
                            <Button
                                variant="outline"
                                type="button"
                                className="margin-top-0-important"
                                onClick={() => {
                                    if (props.onClick) {
                                        props.onClick();
                                    }
                                }}
                            >
                                {t("MetricsAdd")}
                            </Button>
                        </div>
                    )}
                </div>
            ) : (
                <div className="text-center radius-lg padding-3 margin-y-1 border-base border-dashed bg-base-lightest border">
                    <p>{t("MetricsZero")}</p>
                    <div className="text-center">
                        <Button
                            variant="outline"
                            type="button"
                            onClick={() => {
                                if (props.onClick) {
                                    props.onClick();
                                }
                            }}
                        >
                            {t("MetricsAdd")}
                        </Button>
                    </div>
                </div>
            )}
        </div>
    );
}

export default MetricsList;
