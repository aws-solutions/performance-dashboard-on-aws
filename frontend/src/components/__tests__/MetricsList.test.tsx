/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import MetricsList from "../MetricsList";
import { Metric } from "../../models";

test("renders the component", async () => {
    const metrics = [{ title: "temperature", value: 90 }] as Metric[];
    const onClickMock = jest.fn();
    const wrapper = render(
        <MetricsList onClick={onClickMock} metrics={metrics} allowAddMetric={false} />,
    );

    expect(wrapper.container).toMatchSnapshot();
});

test("renders the MetricsAdd button", async () => {
    const metrics = [{ title: "temperature", value: 90 }] as Metric[];
    const onClickMock = jest.fn();
    render(<MetricsList onClick={onClickMock} metrics={metrics} allowAddMetric={true} />);
    const addMetricsButton = screen.getByRole("button", {
        name: "+ Add metric",
    });
    fireEvent.click(addMetricsButton);

    expect(onClickMock).toBeCalled();
});

test("renders zero metrics with  MetricsAdd button", async () => {
    const metrics = [] as Metric[];
    const onClickMock = jest.fn();
    render(<MetricsList onClick={onClickMock} metrics={metrics} allowAddMetric={true} />);
    const addMetricsButton = screen.getByRole("button", {
        name: "+ Add metric",
    });
    fireEvent.click(addMetricsButton);

    expect(screen.getByText("No metrics added yet.")).toBeInTheDocument();
    expect(onClickMock).toBeCalled();
});

test("click on the metric name should trigger onEdit callback", async () => {
    const metrics = [{ title: "temperature", value: 90 }] as Metric[];
    const onEditMock = jest.fn();
    render(
        <MetricsList
            onClick={() => {}}
            onEdit={onEditMock}
            metrics={metrics}
            allowAddMetric={false}
        />,
    );
    const actionsButton = screen.getByRole("button", {
        name: "Edit temperature",
    });
    fireEvent.click(actionsButton);

    expect(onEditMock).toBeCalled();
});

test("click on delete button should trigger onDelete callback", async () => {
    const metrics = [{ title: "temperature", value: 90 }] as Metric[];
    const onDeleteMock = jest.fn();
    render(
        <MetricsList
            onClick={() => {}}
            onDelete={onDeleteMock}
            metrics={metrics}
            allowAddMetric={false}
        />,
    );
    const deleteButton = screen.getByText("Delete");
    fireEvent.click(deleteButton);

    expect(onDeleteMock).toBeCalled();
});

test("click on the up arrow should trigger onMoveUp callback", async () => {
    const metrics = [
        { title: "polution", value: 25 },
        { title: "temperature", value: 90 },
    ] as Metric[];
    const onMoveUpMock = jest.fn();
    render(
        <MetricsList
            onClick={() => {}}
            onMoveUp={onMoveUpMock}
            metrics={metrics}
            allowAddMetric={false}
        />,
    );
    const onMoveUpButton = screen.getByRole("button", {
        name: "Move temperature up",
    });
    fireEvent.click(onMoveUpButton);

    expect(onMoveUpMock).toBeCalled();
});

test("click on the up arrow should trigger onMoveUp callback", async () => {
    const metrics = [
        { title: "polution", value: 25 },
        { title: "temperature", value: 90 },
    ] as Metric[];
    const onMoveDownMock = jest.fn();
    render(
        <MetricsList
            onClick={() => {}}
            onMoveDown={onMoveDownMock}
            metrics={metrics}
            allowAddMetric={false}
        />,
    );
    const onMoveUpButton = screen.getByRole("button", {
        name: "Move polution down",
    });
    fireEvent.click(onMoveUpButton);

    expect(onMoveDownMock).toBeCalled();
});
