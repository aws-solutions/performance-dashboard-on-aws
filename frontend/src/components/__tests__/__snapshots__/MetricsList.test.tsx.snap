// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<div>
  <div
    aria-label="Metrics"
    class="display-block"
    role="group"
  >
    <label
      class="margin-bottom-0 margin-top-2 usa-label text-bold"
    >
      Metrics
    </label>
    <p
      class="margin-top-2px usa-hint"
    >
      Add metrics here and reorder them. Multiple metrics will automatically scale to fit and wrap according to the order.
    </p>
    <div
      class="usa-checkbox margin-bottom-2"
    >
      <input
        class="usa-checkbox__input"
        id="oneMetricPerRow"
        name="oneMetricPerRow"
        type="checkbox"
      />
      <label
        class="usa-checkbox__label"
        for="oneMetricPerRow"
      >
        One metric per row
      </label>
    </div>
    <div>
      <div
        class="padding-top-2 padding-bottom-2"
        data-rbd-droppable-context-id="0"
        data-rbd-droppable-id="metrics-list"
        role="list"
      >
        <div
          aria-label="1"
          class="border-base-lighter border-1px shadow-1 z-200 radius-lg bg-white grid-row margin-y-1"
          data-rbd-draggable-context-id="0"
          data-rbd-draggable-id="0"
          role="listitem"
        >
          <div
            class="grid-row grid-col flex-2 padding-1"
          >
            <div
              aria-describedby="rbd-hidden-text-0-hidden-text-0"
              class="text-base-darker grid-col flex-3 text-center display-flex flex-align-center flex-justify-center margin-left-1"
              data-rbd-drag-handle-context-id="0"
              data-rbd-drag-handle-draggable-id="0"
              draggable="false"
              role="button"
              tabindex="0"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-grip-vertical fa-w-10 fa-xs "
                data-icon="grip-vertical"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 320 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M96 32H32C14.33 32 0 46.33 0 64v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32V64c0-17.67-14.33-32-32-32zm0 160H32c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32v-64c0-17.67-14.33-32-32-32zm0 160H32c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32v-64c0-17.67-14.33-32-32-32zM288 32h-64c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32V64c0-17.67-14.33-32-32-32zm0 160h-64c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32v-64c0-17.67-14.33-32-32-32zm0 160h-64c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32v-64c0-17.67-14.33-32-32-32z"
                  fill="currentColor"
                />
              </svg>
            </div>
            <div
              class="grid-col flex-6 text-center display-flex flex-align-center flex-justify-center font-sans-md margin-left-1"
            >
              1
            </div>
            <div
              class="grid-col flex-4 grid-row flex-column text-center margin-left-1 margin-right-1"
            >
              <div
                class="grid-col flex-6"
              >
                <br />
              </div>
              <div
                class="grid-col flex-6"
              >
                <br />
              </div>
            </div>
          </div>
          <div
            class="border-base-lighter border-left"
          />
          <div
            class="grid-col flex-10 grid-row margin-y-1"
          >
            <div
              class="grid-col flex-11 usa-tooltip"
              data-position="bottom"
              title="temperature"
            >
              <div
                class="margin-left-1 text-no-wrap overflow-hidden text-overflow-ellipsis"
              >
                <button
                  aria-label="Edit temperature"
                  class="usa-button usa-button--unstyled margin-left-1 margin-top-0-important text-base-dark hover:text-base-darker active:text-base-darkest text-bold"
                  type="button"
                >
                  temperature
                </button>
              </div>
            </div>
            <div
              class="grid-col flex-1 text-right"
            >
              <button
                aria-controls="menu--1"
                aria-haspopup="true"
                aria-label="Actions"
                class="usa-button usa-button--unstyled margin-1"
                data-reach-menu-button=""
                id="menu-button--menu--1"
                type="button"
              >
                
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-ellipsis-v fa-w-6 margin-left-1"
                  data-icon="ellipsis-v"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 192 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
